Source/WTF/wtf/ThreadSafeWeakHashSet.h

```diff
@@ -0,0 +1,174 @@
+        using value_type = T;
+        using pointer = const value_type*;
+        using reference = const value_type&;
+
+        T* get() const { return m_strongReferences[m_position].ptr(); }
+        T& operator*() const { return *get(); }
+        T* operator->() const { return get(); }
+
+        const_iterator& operator++()
+        {
+            m_position++;
+            return *this;
+        }
+
+        bool operator==(const const_iterator& other) const
+        {
+            // Not a complete comparison, but only used to compare an iterator against the end.
+            ASSERT_UNUSED(other, !other.m_strongReferences.size());
+            return m_position >= m_strongReferences.size();
+        }
+
+    private:
+        template<typename> friend class ThreadSafeWeakHashSet;
+        const_iterator(Vector<Ref<T>>&& strongReferences)
+            : m_strongReferences(WTFMove(strongReferences)) { }
+
+        const Vector<Ref<T>> m_strongReferences;
+        size_t m_position { 0 };
+    };
+
+    const_iterator begin() const
+    {
+        Vector<Ref<T>> strongReferences;
+        Locker locker { m_lock };
+        m_set.removeIf([&] (auto& controlBlock) {
+            if (auto refPtr = controlBlock->makeStrongReferenceIfPossible()) {
+                strongReferences.append(refPtr.releaseNonNull());
+                return false;
+            }
+            return true;
+        });
+        m_operationCountSinceLastCleanup = 0;
+        return { WTFMove(strongReferences) };
+    }
+
+    const_iterator end() const
+    {
+        return { { } };
+    }
+
+    template<typename U, std::enable_if_t<std::is_convertible_v<U*, T*>>* = nullptr>
+    typename HashSet<Ref<ThreadSafeWeakPtrControlBlock<T>>>::AddResult add(const U& value)
+    {
+        amortizedCleanupIfNeeded();
+        RELEASE_ASSERT_WITH_SECURITY_IMPLICATION(!value.m_controlBlock.objectHasBeenDeleted());
+        Locker locker { m_lock };
+        return m_set.add(reinterpret_cast<ThreadSafeWeakPtrControlBlock<T>&>(value.m_controlBlock));
+    }
+
+    template<typename U, std::enable_if_t<std::is_convertible_v<U*, T*>>* = nullptr>
+    bool remove(const U& value)
+    {
+        amortizedCleanupIfNeeded();
+        RELEASE_ASSERT_WITH_SECURITY_IMPLICATION(!value.m_controlBlock.objectHasBeenDeleted());
+        Locker locker { m_lock };
+        return m_set.remove(reinterpret_cast<ThreadSafeWeakPtrControlBlock<T>&>(value.m_controlBlock));
+    }
+
+    void clear()
+    {
+        Locker locker { m_lock };
+        m_set.clear();
+        m_operationCountSinceLastCleanup = 0;
+    }
+
+    template<typename U, std::enable_if_t<std::is_convertible_v<U*, T*>>* = nullptr>
+    bool contains(const U& value) const
+    {
+        Locker locker { m_lock };
+        ++m_operationCountSinceLastCleanup;
+        return m_set.contains(value.m_controlBlock);
+    }
+
+    bool computesEmpty() const { return begin() == end(); }
```

Comment:
Given that begin() allocates a vector, this is not an efficient implementation for figuring out if the container is empty.